modules/packages/LinearAlgebra.chpl

```diff
@@ -0,0 +1,709 @@
+}
+
+
+/* Return a matrix (2D array) with domain and values of ``A`` */
+proc Matrix(A: [?Dom] ?eltType) where Dom.rank == 2 {
+  var M: [Dom] eltType = A;
+  return M;
+}
+
+
+pragma "no doc"
+proc Matrix(Arrays...?n) {
+  type eltType = Arrays(1).eltType;
+  return Matrix((...Arrays), eltType=eltType);
+}
+
+
+/*
+    Return a matrix (2D array), given 2 or more vectors. Vectors are
+    concatenated such that the ``ith`` vector corresponds to the matrix slice:
+    ``A[i, ..]``
+
+    If `type` is omitted, it will be inferred from the first argument
+*/
+proc Matrix(const Arrays...?n, type eltType) {
+  // TODO -- assert all array domains are same length
+  //         Can this be done via type query?
+
+  if Arrays(1).domain.rank != 1 then compilerError("Matrix() expected 1D arrays");
+
+  const dim2 = 0..#Arrays(1).domain.dim(1).length,
+        dim1 = 0..#n;
+
+  var M: [{dim1, dim2}] eltType;
+
+  for i in dim1 do
+    M[i, ..] = Arrays(i+1)[..]: eltType;
+
+  return M;
+}
+
+
+/* Return an identity matrix over domain ``{0..#m, 0..#n}`` */
+proc eye(m, n, type eltType=real) {
+  var A: [{0..#m, 0..#n}] eltType;
+  const diagonal = min(m, n);
+  for i in 0..#diagonal do A[i, i] = 1: eltType;
+  return A;
+}
+
+
+/* Return an identity matrix over domain ``Dom`` */
+proc eye(Dom: domain(2), type eltType=real) {
+  var A: [Dom] eltType;
+  const diagonal = min(Dom.shape(1), Dom.shape(2));
+  return eye(Dom.shape(1), Dom.shape(2), eltType=eltType);
+}
+
+
+//
+// Matrix Operations
+//
+
+pragma "no doc"
+/* Transpose vector (no-op) */
+inline proc transpose(A: [?Dom]) where Dom.rank == 1 {
+  return A;
+}
+
+
+pragma "no doc"
+/* Syntactic sugar for transpose(Vector) */
+proc _array.T where this.domain.rank == 1 { return transpose(this); }
+
+
+/* Transpose vector or matrix.
+
+   .. note::
+
+      Since row vectors and columns vectors are indistinguishable, passing
+      a vector to this function will return that vector unchanged
+
+*/
+proc transpose(A: [?Dom] ?eltType) where Dom.rank == 2 {
+  if !isBLASType(eltType) then
+    return _transpose(A);
+  else if Dom.shape(1) == 1 then
+    return reshape(A, {Dom.dim(2), Dom.dim(1)});
+  else if Dom.shape(2) == 1 then
+    return reshape(A, {Dom.dim(2), Dom.dim(1)});
+  else {
+    var B: [Dom] eltType = eye(Dom, eltType=eltType);
+    var C: [{Dom.dim(2), Dom.dim(1)}] eltType;
+    // TODO -- use native algorithm (BLAS not necessary or memory-efficient)
+    gemm(A, B, C, 1:eltType, 0:eltType, opA=Op.T);
+    return C;
+  }
+}
+
+
+/* Transpose vector or matrix */
+proc _array.T where this.domain.rank == 2 { return transpose(this); }
+
+
+/* Transpose matrix with native implementation */
+pragma "no doc"
+private proc _transpose(A: [?Dom] ?eltType) {
+  var C: [{Dom.dim(2), Dom.dim(1)}] eltType;
+
+  // naive algorithm
+  forall (i, j) in Dom do
+    C[j, i] = A[i, j];
+
+  return C;
+}
+
+
+/* Add matrices, maintaining dimensions */
+proc matPlus(A: [?Adom] ?eltType, B: [?Bdom] eltType) {
```

Comment:
I think these (matPlus, matMinus, etc.) should have where-clauses or compiler asserts ensuring that the rank is two (or maybe one as well).